app/controllers/candidate_interface/sign_in_controller.rb

```diff
@@ -12,38 +12,48 @@ def create
       SignInCandidate.new(candidate_params[:email_address], self).call
     end
 
+    # This is where users land after clicking the magic link in their email. Because
+    # some email clients preload links in emails, this is an extra step where
+    # they click a button to confirm the sign in.
     def confirm_authentication
-      candidate = retrieve_candidate_by_token || retrieve_candidate_by_user_id
+      authentication_token = AuthenticationToken.find_by_hashed_token(
+        authenticable_type: 'Candidate',
+        raw_token: params[:token],
+      )
 
-      redirect_to(action: :new) and return if candidate.nil?
-
-      if FindCandidateByToken.token_not_expired?(candidate)
+      if authentication_token && authentication_token.still_valid?
         render 'confirm_authentication'
-      else
+      elsif params[:u]
+        # If the token is invalid or expired, redirect the user to a page where
+        # they can request a new sign in email.
         redirect_to candidate_interface_expired_sign_in_path(u: params[:u], path: params[:path])
+      else
+        redirect_to(action: :new)
       end
     end
 
+    # After they click the confirm button, actually do the user sign in.
     def authenticate
-      candidate = retrieve_candidate_by_token
-      token_not_expired = FindCandidateByToken.token_not_expired?(candidate)
+      authentication_token = AuthenticationToken.find_by_hashed_token(
+        authenticable_type: 'Candidate',
+        raw_token: params[:token],
+      )
 
-      if candidate.nil?
-        candidate = retrieve_candidate_by_user_id
-      end
-
-      redirect_to(action: :new) and return if candidate.nil?
+      redirect_to(action: :new) and return if authentication_token.nil?
 
-      if token_not_expired
+      if authentication_token && authentication_token.still_valid?
+        candidate = authentication_token.authenticable
         flash[:success] = t('apply_from_find.account_created_message') if candidate.last_signed_in_at.nil?
         sign_in(candidate, scope: :candidate)
-        add_identity_to_log candidate.id
-        candidate.update_sign_in_fields!
+        add_identity_to_log(candidate.id)
+        candidate.update!(last_signed_in_at: Time.zone.now)
+        authentication_token.destroy!
```

Comment:
I'm wondering if it's worth just having an extra column for this model like 'used' which we set to true after candidate signs in, rather than deleting it. That way we can run queries really easily without having to use audits. For example, what % of links get clicked would just be:

`ApplicationToken.where(used: true).where(authenticable_type: 'Candidate') / ApplicationToken.token.where(authenticable_type: 'Candidate') `

Reply:
Yes, good idea! I'll make a follow up PR for this because it will require the system tests to be updated, and I don't want to change both behaviour and tests in the same PR (so we can be sure this PR doesn't change core bahaviour).